### PR TITLE
Add Beam Search decoding algorithm

### PR DESCRIPTION
**Pull request by artek0chumak**
Add the first version of the Beam search algorithm. Not efficient and can contain bugs.
TODO:
- [x] Tests and comparisons with hf's beam search
- [x] Filter out constraints
- [x] Check multibatches scenario.
- [ ] Add new tokenizer behaviour.

### CODE REVIEW

File: src/client/remote_generation.py [hunk 1]
```diff
@@ -1,10 +1,18 @@
 from typing import List, Optional
 
 import torch
-import torch.nn.functional as F
+from hivemind import get_logger
```
Comment (borzunov) [inline src/client/remote_generation.py:4]: nit:
```suggestion
from hivemind.utils.logging import get_logger
```

File: src/client/remote_generation.py [hunk 2]
```diff
@@ -78,24 +93,43 @@ def generate(
...
+                decoding_algorithm = BeamSearchAlgorithm(num_beams, batch_size=batch_size)
             else:
                 decoding_algorithm = GreedyAlgorithm()
 
+        if num_beams > 1:
+            inputs = torch.cat([inputs] * num_beams, dim=0)
+            if batch_size > 1:
+                # TODO: resolve padding problem
+                logger.warning(
+                    f"You set batch_size {batch_size} within beam search generation. Be carefull, results on sequences with different length may be padded wrong way"
```
Comment (borzunov) [inline src/client/remote_generation.py:111]:
```suggestion
                    f"You set batch_size {batch_size} within beam search generation. Be careful, results on sequences with different length may be padded wrong way"
```

File: src/client/remote_generation.py [hunk 3]
```diff
@@ -78,24 +93,43 @@ def generate(
...
+                )
+
+        if num_return_sequences is None:
+            num_return_sequences = 1
+
+        assert num_return_sequences <= num_beams, (
+            f"You want more sequences that beam will have."
```
Comment (borzunov) [inline src/client/remote_generation.py:118]:
```suggestion
            f"You want more sequences than the beam has."
```

File: tests/test_full_model.py [hunk 4]
```diff
@@ -89,3 +90,30 @@ def test_greedy_generation(max_new_tokens=4):
...
+        num_beams=num_beams,
+        device=inputs.device,
+        length_penalty=0,
+        do_early_stopping=False,
+    )
+    hf_inputs = tokenizer([text] * 2, return_tensors="pt")["input_ids"]
+    hf_outputs = BloomForCausalLM.beam_search(
+        model, input_ids=hf_inputs, max_length=inputs.size(1) + max_new_tokens, beam_scorer=beam_scorer
+    )
+    assert torch.allclose(remote_outputs, hf_outputs), "Beam search are not identical to HF"
```
Comment (borzunov) [inline tests/test_full_model.py:119]:
```suggestion
    assert torch.allclose(remote_outputs, hf_outputs), "Beam search results are not identical to HF"
```

**Review by borzunov — APPROVED** [review]

**Review by justheuristic — APPROVED** [review]
Good job! i have no significant comments

File: src/client/remote_generation.py [hunk 5]
```diff
@@ -117,19 +151,34 @@ def generate(
                 for constraint in constraints:
                     lm_logits = constraint(last_token_id, lm_logits, hypo_ids)
                 last_token_id, hypo_ids = decoding_algorithm(lm_logits)
-                if seq_idx < inputs.size(1):  # TODO: why is it not a constraint?
+
+                # If samples have padded, so changes only them
```
Comment (justheuristic) [inline src/client/remote_generation.py:155]: Nit: "samples have padded" sounds a bit confusing (iirc, "to pad" means to apply to something else, not to receive padding)

Suggestion: "If some samples were padded, change only these samples"?
